### ISSUE TITLE
[parquet-protobuf] stack overflow when work with well known json type

### ISSUE THREAD

**Issue by asfimport** [apache/parquet-java#2397]
Writing following protobuf message as parquet file is not possible: 
```java

syntax = "proto3";

import "google/protobuf/struct.proto";

package test;
option java_outer_classname = "CustomMessage";

message TestMessage {
    map<string, google.protobuf.ListValue> data = 1;
} 
```
Protobuf introduced "well known json type" such like [ListValue](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#listvalue) to work around json schema conversion. 

However writing above messages traps parquet writer into an infinite loop due to the "general type" support in protobuf. Current implementation will keep referencing 6 possible types defined in protobuf (null, bool, number, string, struct, list) and entering infinite loop when referencing "struct".
```java

java.lang.StackOverflowErrorjava.lang.StackOverflowError at java.base/java.util.Arrays$ArrayItr.<init>(Arrays.java:4418) at java.base/java.util.Arrays$ArrayList.iterator(Arrays.java:4410) at java.base/java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1044) at java.base/java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1043) at org.apache.parquet.proto.ProtoSchemaConverter.convertFields(ProtoSchemaConverter.java:64) at org.apache.parquet.proto.ProtoSchemaConverter.addField(ProtoSchemaConverter.java:96) at org.apache.parquet.proto.ProtoSchemaConverter.convertFields(ProtoSchemaConverter.java:66) at org.apache.parquet.proto.ProtoSchemaConverter.addField(ProtoSchemaConverter.java:96) at org.apache.parquet.proto.ProtoSchemaConverter.convertFields(ProtoSchemaConverter.java:66) at org.apache.parquet.proto.ProtoSchemaConverter.addField(ProtoSchemaConverter.java:96) at org.apache.parquet.proto.ProtoSchemaConverter.convertFields(ProtoSchemaConverter.java:66) at org.apache.parquet.proto.ProtoSchemaConverter.addField(ProtoSchemaConverter.java:96) at org.apache.parquet.proto.ProtoSchemaConverter.convertFields(ProtoSchemaConverter.java:66) at org.apache.parquet.proto.ProtoSchemaConverter.addField(ProtoSchemaConverter.java:96) 
```

**Reporter**: [Lawrence He](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=holyshipt)

<sub>**Note**: *This issue was originally created as [PARQUET-1711](https://issues.apache.org/jira/browse/PARQUET-1711). Please see the [migration documentation](https://issues.apache.org/jira/browse/PARQUET-2502) for further details.*</sub>

**Comment by asfimport** [comment]
[Lawrence He](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-16989291):
We also ran into this issue with latest master parquet build:
```java

java.lang.StackOverflowErrorjava.lang.StackOverflowError at org.apache.parquet.schema.Types$BasePrimitiveBuilder$1.checkBinaryPrimitiveType(Types.java:561) at org.apache.parquet.schema.Types$BasePrimitiveBuilder$1.visit(Types.java:448) at org.apache.parquet.schema.LogicalTypeAnnotation$StringLogicalTypeAnnotation.accept(LogicalTypeAnnotation.java:303) at org.apache.parquet.schema.Types$BasePrimitiveBuilder.build(Types.java:445) at org.apache.parquet.schema.Types$BasePrimitiveBuilder.build(Types.java:336) at org.apache.parquet.schema.Types$Builder.named(Types.java:314) at org.apache.parquet.proto.ProtoSchemaConverter.convertFields(ProtoSchemaConverter.java:83) at org.apache.parquet.proto.ProtoSchemaConverter.addMessageField(ProtoSchemaConverter.java:147) at org.apache.parquet.proto.ProtoSchemaConverter.addField(ProtoSchemaConverter.java:100) at org.apache.parquet.proto.ProtoSchemaConverter.convertFields(ProtoSchemaConverter.java:81) at org.apache.parquet.proto.ProtoSchemaConverter.addMessageField(ProtoSchemaConverter.java:147) at org.apache.parquet.proto.ProtoSchemaConverter.addField(ProtoSchemaConverter.java:100) at org.apache.parquet.proto.ProtoSchemaConverter.convertFields(ProtoSchemaConverter.java:81) at org.apache.parquet.proto.ProtoSchemaConverter.addMessageField(ProtoSchemaConverter.java:147) at org.apache.parquet.proto.ProtoSchemaConverter.addField(ProtoSchemaConverter.java:100) at org.apache.parquet.proto.ProtoSchemaConverter.convertFields(ProtoSchemaConverter.java:81) at org.apache.parquet.proto.ProtoSchemaConverter.addMessageField(ProtoSchemaConverter.java:147) 
```

**Comment by asfimport** [comment]
[J Y](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17542604) / @jinyius:
i believe google's [Struct](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/struct.proto) also runs afoul here, too.

any new progress here?

**Comment by asfimport** [comment]
[Aaron Blake Niskode-Dossett](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17542642) / @dossett:
That is a very interesting issue.  I'm still learning a lot about parquet, but I'm not sure parquet has a notion of a recursive schema the way protobuf does here.  In other words, I'm not sure how parquet **could** handle this (although it should not be a stack overflow!)

**Comment by asfimport** [comment]
[Micah Kornfield](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17543672) / @emkornfield:
the way one could handle this is allow users to specify a recursion limit, and expand the schema to that limit (this would need to be stored in metadata).  For protos that end up deeper then that level you either error on write, or have the leaf levels of recursion store serialized proto bytes.

**Comment by asfimport** [comment]
[Reeni Mathew](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17545732):
@emkornfield Can we expect a fix any time soon?

Hit by the same issue when Schema contains a Self Reference.

**Comment by asfimport** [comment]
[J Y](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17552396) / @jinyius:
i'd be ok of that approach: a proto option annotation to limit the recursion limit, then failing over to treat it as proto bytes.  if the recursion limit is omitted/missing, then just treat the recursive definition as bytes after the first occurrence.

forgive me if this is a naive question, but what's the difficulty in supporting "typing" properly to handle recursive nesting?

 

PARQUET-129 is very much related/the same issue...

**Comment by asfimport** [comment]
[Micah Kornfield](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17570331) / @emkornfield:
> @emkornfield  Can we expect a fix any time soon?
I was not planning on working on this, I don't have a lot of expertise on parquet-mr, but I can help review if someone wants to try this approach.

**Comment by asfimport** [comment]
[Gidon Gershinsky](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17602127) / @ggershinsky:
Hi to all on this Jira. Looks like we have a number of alternative solutions to this problem today, 
<https://github.com/apache/parquet-mr/pull/995>

<https://github.com/apache/parquet-mr/pull/445>

<https://github.com/apache/parquet-mr/pull/988>

Can you take a look and provide your opinion on them?

**Comment by asfimport** [comment]
[Gidon Gershinsky](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17605098) / @ggershinsky:
@emkornfield what do you think about these 3 alternatives?

**Comment by asfimport** [comment]
[J Y](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17608085) / @jinyius:
ping.. the pr i wrote is the byte encoding version that [~emkornfield@gmail.com] suggested in https://issues.apache.org/jira/browse/PARQUET-1711?focusedCommentId=17543672&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17543672

**Comment by asfimport** [comment]
[J Y](https://issues.apache.org/jira/browse/PARQUET-1711?#comment-17638109) / @jinyius:
this should be resolved after #995 now.
